Userland/Libraries/LibCore/Socket.cpp

```diff
@@ -208,13 +208,18 @@ ErrorOr<NonnullOwnPtr<TCPSocket>> TCPSocket::connect(SocketAddress const& addres
     return socket;
 }
 
+// The socket must already be connected
 ErrorOr<NonnullOwnPtr<TCPSocket>> TCPSocket::adopt_fd(int fd)
 {
     if (fd < 0) {
         return Error::from_errno(EBADF);
     }
 
-    auto socket = TRY(adopt_nonnull_own_or_enomem(new (nothrow) TCPSocket()));
+    struct sockaddr_in addr;
+    socklen_t addr_size = sizeof(struct sockaddr_in);
+    TRY(System::getpeername(fd, (struct sockaddr*)&addr, &addr_size));
```

Comment:
This looks very c-y
in c++ you do not need to say that the type is indeed a struct
also please prefer `static_cast` and the like

Also `sockaddr` and `sockaddr_in` don't seem to be compatible in layout, but i do not know the socket interface, so this could be expected, iff so ignore this last statement

Reply:
I fixed the cast.
As for using `sockaddr` and `sockaddr_in` this is on purpose: https://stackoverflow.com/questions/21099041/why-do-we-cast-sockaddr-in-to-sockaddr-when-calling-bind